metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -0,0 +1,243 @@
+func NewContainerMetadataEnricher(
+	base mb.BaseMetricSet,
+	nodeScope bool) Enricher {
+
+	watcher, err := GetWatcher(base, &kubernetes.Pod{}, nodeScope)
+	if err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	if watcher == nil {
+		logp.Info("Kubernetes metricset enriching is disabled")
+		return nilEnricher
+	}
+
+	metaConfig := kubernetes.MetaGeneratorConfig{}
+	if err := base.Module().UnpackConfig(&metaConfig); err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	metaGen := kubernetes.NewMetaGeneratorFromConfig(&metaConfig)
+	enricher := buildMetadataEnricher(watcher,
+		// update
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			pod := r.(*kubernetes.Pod)
+			meta := metaGen.ResourceMetadata(r)
+			for _, container := range append(pod.GetSpec().GetContainers(), pod.GetSpec().GetInitContainers()...) {
+				id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName(), container.GetName())
+				m[id] = meta
+			}
+		},
+		// delete
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			pod := r.(*kubernetes.Pod)
+			for _, container := range append(pod.GetSpec().GetContainers(), pod.GetSpec().GetInitContainers()...) {
+				id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName(), container.GetName())
+				delete(m, id)
+			}
+		},
+		// index
+		func(e common.MapStr) string {
+			return join(getString(e, mb.ModuleDataKey+".namespace"), getString(e, mb.ModuleDataKey+".pod.name"), getString(e, "name"))
+		},
+	)
+
+	return enricher
+}
+
+func getString(m common.MapStr, key string) string {
+	val, err := m.GetValue(key)
+	if err != nil {
+		return ""
+	}
+
+	str, _ := val.(string)
+	return str
+}
+
+func join(fields ...string) string {
+	return strings.Join(fields, ":")
+}
+
+func buildMetadataEnricher(
+	watcher kubernetes.Watcher,
+	update func(map[string]common.MapStr, kubernetes.Resource),
+	delete func(map[string]common.MapStr, kubernetes.Resource),
+	index func(e common.MapStr) string) Enricher {
+
+	enricher := enricher{
+		metadata: map[string]common.MapStr{},
+		index:    index,
+		watcher:  watcher,
+	}
+
+	watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
+		AddFunc: func(obj kubernetes.Resource) {
+			enricher.Lock()
+			defer enricher.Unlock()
+			update(enricher.metadata, obj)
+		},
+		UpdateFunc: func(obj kubernetes.Resource) {
+			enricher.Lock()
+			defer enricher.Unlock()
+			update(enricher.metadata, obj)
+		},
+		DeleteFunc: func(obj kubernetes.Resource) {
+			enricher.Lock()
+			defer enricher.Unlock()
+			delete(enricher.metadata, obj)
+		},
+	})
+
+	return &enricher
+}
+
+func (m *enricher) Start() {
+	if m == nilEnricher {
+		return
+	}
+
+	if !m.watcherStarted {
+		err := m.watcher.Start()
+		if err != nil {
+			logp.Warn("Error starting Kubernetes watcher: %s", err)
+		}
+		m.watcherStarted = true
+	}
+}
```

Comment:
And a `Stop()` to release the watcher resources?

Reply:
Added, thank you for bringing this in, it was in my TODO :+1: